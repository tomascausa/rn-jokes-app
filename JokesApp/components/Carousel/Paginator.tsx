import React from 'react';
import {Animated, StyleSheet, Dimensions} from 'react-native';

const {width} = Dimensions.get('window');

const Paginator = ({scrollX, itemsQty}: {scrollX: any; itemsQty: number}) => {
    return (
        <Animated.View style={styles.stepperContainer}>
            {Array(itemsQty < 4 ? itemsQty : 4)
                .fill(1)
                .map((_, i) => {
                    const inputRange = [
                        (i - 1) * width,
                        i * width,
                        (i + 1) * width,
                    ];

                    const opacity = scrollX.interpolate({
                        inputRange,
                        outputRange: [0.2, 1, 0.2],
                        extrapolate: 'clamp',
                    });

                    return (
                        <Animated.View
                            style={[styles.step, {opacity}]}
                            key={i}
                        />
                    );
                })}
        </Animated.View>
    );
};

const styles = StyleSheet.create({
    stepperContainer: {
        flexDirection: 'row',
        justifyContent: 'center',
        marginTop: 35,
    },
    step: {
        width: 50,
        height: 5,
        marginHorizontal: 8,
        backgroundColor: 'black',
        borderRadius: 8,
    },
});

export default Paginator;
