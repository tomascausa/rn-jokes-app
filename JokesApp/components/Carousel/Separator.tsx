import React from 'react';
import {View, StyleSheet} from 'react-native';

const Separator: React.FC = () => {
    return <View style={styles.container} />;
};

const styles = StyleSheet.create({
    container: {
        width: 10,
        height: 10,
    },
});

export default Separator;
