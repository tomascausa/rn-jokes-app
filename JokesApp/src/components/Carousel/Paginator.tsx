import React from 'react';
import {Animated, StyleSheet} from 'react-native';

interface PaginatorProps {
    currentIndex: number;
}

const Paginator = ({currentIndex}: PaginatorProps) => {
    return (
        <Animated.View style={styles.stepperContainer}>
            <>
                {Array(4)
                    .fill(1)
                    .map((_, i) => {
                        return (
                            <Animated.View
                                style={[
                                    styles.step,
                                    i === currentIndex
                                        ? styles.activeStep
                                        : null,
                                ]}
                                key={i}
                            />
                        );
                    })}
            </>
        </Animated.View>
    );
};

const styles = StyleSheet.create({
    stepperContainer: {
        flexDirection: 'row',
        justifyContent: 'center',
        marginTop: 10,
    },
    step: {
        width: 50,
        height: 5,
        marginHorizontal: 8,
        backgroundColor: 'black',
        borderRadius: 8,
        opacity: 0.3,
    },
    activeStep: {
        opacity: 1,
    },
});

export default Paginator;
