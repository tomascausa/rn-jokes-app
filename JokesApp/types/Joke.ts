export type Joke = {
    id: string;
    message: string;
    author?: string;
};
