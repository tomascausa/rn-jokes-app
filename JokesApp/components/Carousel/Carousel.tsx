import React, {FC} from 'react';
import {View, Animated, StyleSheet, Dimensions} from 'react-native';
import Paginator from './Paginator';

interface Props {
    items: any[];
    renderItem: any;
    onIndexUpdate?: (number: number) => void;
}

const {width} = Dimensions.get('window');

const Carousel: FC<Props> = ({items, renderItem, onIndexUpdate}) => {
    const scrollX = React.useRef(new Animated.Value(0)).current;

    return (
        <View style={styles.carouselContainer}>
            <Animated.FlatList
                data={items}
                renderItem={renderItem.bind(this, scrollX)}
                horizontal
                showsHorizontalScrollIndicator={false}
                scrollEventThrottle={32}
                pagingEnabled
                onScroll={Animated.event(
                    [{nativeEvent: {contentOffset: {x: scrollX}}}],
                    {useNativeDriver: false},
                )}
                onMomentumScrollEnd={event => {
                    if (onIndexUpdate) {
                        let index = Math.ceil(
                            event.nativeEvent.contentOffset.x / width,
                        );
                        onIndexUpdate(index);
                    }
                }}
            />
            <Paginator scrollX={scrollX} itemsQty={items.length} />
        </View>
    );
};

const styles = StyleSheet.create({
    carouselContainer: {
        flex: 1,
    },
});

export default Carousel;
