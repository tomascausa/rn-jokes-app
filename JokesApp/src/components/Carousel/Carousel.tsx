import React, {useRef} from 'react';
import {View, Animated, StyleSheet, ViewToken} from 'react-native';
import Paginator from './Paginator';

interface CarouselProps {
    items: any[];
    renderItem: any;
    isLoading: boolean;
    indicatorIndex: number;
    onIndexUpdate?: (number: number) => void;
    onFetchMore?: () => void;
}

const Carousel = ({
    items,
    renderItem,
    isLoading = false,
    indicatorIndex,
    onIndexUpdate,
    onFetchMore,
}: CarouselProps) => {
    const scrollX = React.useRef(new Animated.Value(0)).current;

    const onEndReached = () => {
        if (onFetchMore) {
            onFetchMore();
        }
    };

    const onViewableItemsChanged = React.useRef(
        ({changed}: {changed: ViewToken[]}) => {
            if (changed && changed.length > 0 && onIndexUpdate) {
                onIndexUpdate(changed[0].index || 0);
            }
        },
    );

    const viewabilityConfig = useRef({
        itemVisiblePercentThreshold: 90,
        waitForInteraction: true,
        minimumViewTime: 5,
    });

    return (
        <View
            style={[
                styles.carouselContainer,
                isLoading ? styles.carouselLoading : null,
            ]}>
            <Animated.FlatList
                testID={'flatList'}
                data={items}
                renderItem={renderItem.bind(this, scrollX)}
                horizontal
                showsHorizontalScrollIndicator={false}
                scrollEventThrottle={64}
                pagingEnabled
                keyExtractor={(_, index) => index.toString()}
                onScroll={Animated.event(
                    [{nativeEvent: {contentOffset: {x: scrollX}}}],
                    {useNativeDriver: false},
                )}
                onViewableItemsChanged={onViewableItemsChanged.current}
                viewabilityConfig={viewabilityConfig.current}
                onEndReached={onEndReached.bind(this)}
            />
            {!isLoading && <Paginator currentIndex={indicatorIndex} />}
        </View>
    );
};

const styles = StyleSheet.create({
    carouselContainer: {
        flex: 1,
    },
    carouselLoading: {
        opacity: 0.3,
    },
});

export default Carousel;
