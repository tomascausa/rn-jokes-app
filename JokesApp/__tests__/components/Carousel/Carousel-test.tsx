import 'react-native';
import React from 'react';
import Carousel from '../../../src/components/Carousel/Carousel';
import {Joke} from '../../../src/types/Joke';
import renderer from 'react-test-renderer';

const jokes: Joke[] = [
    {id: 'j1', joke: 'Joke 1'},
    {id: 'j2', joke: 'Joke 2'},
];

describe('<Carousel />', () => {
    it('renders correctly', () => {
        const json = renderer
            .create(<Carousel items={jokes} renderItem={jest.fn()} />)
            .toJSON();
        expect(json).toMatchSnapshot();
    });
});
