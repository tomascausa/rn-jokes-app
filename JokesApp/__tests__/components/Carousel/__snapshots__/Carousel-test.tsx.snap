// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    collapsable={false}
    data={
      Array [
        Object {
          "id": "1",
          "joke": "test",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    pagingEnabled={true}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={32}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={Object {}}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      />
    </View>
  </RCTScrollView>
  <View
    collapsable={false}
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": 10,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "black",
          "borderRadius": 8,
          "height": 5,
          "marginHorizontal": 8,
          "opacity": 1,
          "width": 50,
        }
      }
    />
  </View>
</View>
`;
